JustSaying/AwsTools/MessageHandling/MessageDispatcher.cs

```diff
@@ -112,7 +117,7 @@ public async Task DispatchMessage(SQSMessage message, CancellationToken cancella
             }
             finally
             {
-                if (!handlingSucceeded && _messageBackoffStrategy != null)
+                if (!handlingSucceeded && _messageBackoffStrategy != null && !(lastException is INonRetryable))
```

Comment:
Maybe an alternative to this would be some sort of strategy that people could implement to determine whether something is retryable or not? That would allow exceptions for types people don't own (or throw) in their own code to be made non-retryable.

Using an interface and/or attributes to mark makes this less generally useful as it's all compile-time based to extend, rather than something that can be controlled at runtime.

For example:

```csharp
public interface IRetryStrategy
{
    bool CanRetry(Exception exception);
}

IRetryStrategy retryStrategy = ...;
if (!handlingSucceeded && _messageBackoffStrategy != null && !retryStrategy.CanRetry(lastException))
{
    // ...
}
```

Reply:
Did think about this and thought that this would be more invasive. It would require to deliver the strategy to the dispatcher and looking at other activity this would cross cut some other IoC and deprecation stories in play. 

I think that in Handler code it is not much effort to handle any non-owned exception and throw one that derives this interface this would normally be more explicit. 

Not sure I follow the distinction between compile and run-time specifically how this effects 'usefulness'. I guess it does deliver some benefits if the strategy gets more context than just the exception, but that does trade flexibly for complexity

Reply:
So to take an extreme example to illustrate the point, if I wanted to not retry on `OutOfMemoryException`, that exception type lives in the framework, so I can't change it to add an attribute or interface, and it's thrown by the runtime, so I can't change how and where it's thrown either.

In the case of using the strategy instead of an interface/attribute, it's possible for an integrator to easily specify this behaviour in their implementation of `IRetryStrategy`.

Reply:
Possibly. And I realise this is an extreme case but, if you can't handle `OutOfMemoryException` in order to throw your own exception, it stands a chance that '**all bets are off**' so your `IRetryStrategy` would not get a look in either. Also I would not condone short circuit for such a Exception as the instance that is OOM is toast but other instances will processes this message. 

In fact I think that it is clearer to handle exceptions in the Handler explicitly and re-throw a purposed error rather than bubble up system exceptions and decide elsewhere what may have caused them and whether or not to defeat the re-drive policy. 

if 

> I can't change how and where it's thrown either

then you should probably let the SQS redrive do its thing. 

So even if the `IRetryStrategy` existed I would still be catching exceptions and re-throwing as specific exception. Then handling the specific exception in the implementation of `IRetryStrategy`. 

I'm not denigrating the idea of a factory, but it would required changes to many classes to deliver it to the `MessageDespatcher` that would involve changes `IMayWantOptionalSettings`, `SqsNotificationListener`,`JustSayingFluently` etc and it looked like there is WIP on these in other tickets. 

I get that once published the Interface would always have to work so if we find a use case later that is not supported it will be a PITA. But it also adds more complexity to the consumer. Normally I would say 'take the first bullet' and see if a real use case presents that a Boolean retry or not does not satisfy, the strategy could be added later if more complex scenarios present. But I understand it is very much your call so I will go with `IRetryStrategy` if you require. Might take a little longer to get back on this as I guess it will fan out the points of change.

Reply:
I just picked OOM as an example of something built in you can't change, not the OOM condition itself.

Regarding the number of changes, we're currently midway doing lots of refactoring and redesign for v7, so wouldn't necessarily be a problem on that front as we're already making breaking changes and this PR wouldn't be part of any 6.x release.

Reply:
Yep sure, I appreciate OOM was just an example. 

The point I was trying to elaborate was I don't think it would be a good idea to handle **any** system errors anyway. There is a lot of opportunity to think that we do know where the exception came from and assume that makes the message non-retryable. So IMHO the strategy would be more complex to the consumer and be likely to lead to inadvertent dropping of messages to the error queue as opposed to explicitly having to consciously emit a specific exception in the handler in response a specific exception at a known place. 

So perhaps a compromise would be to go with the `IRetryStrategy` but provide a default that does what I have now?

Reply:
Sorry for the delayed response, yes I think a default implementation that does that would be a reasonable compromise.